File: median.exs
defmodule Median do

    def find([h1|t1], [h2|t2]) do
        case h1 < h2 do
             true -> find(t1, [h2|t2], [h1])
             _ -> find([h1|t1], t2, [h2])
        end
    end

    def find(l, []) do
        med(l)
    end

    def find([], l) do
        med(l)
    end

    defp find([], [], result_list) do
        med(result_list)
    end

    defp find([], [h|t], result_list) do
        find([], t, [h|result_list])
    end
    
    defp find([h|t], [], result_list) do
        find(t, [], [h|result_list])
    end

    defp find([h1|t1], [h2|t2], result_list) when h1 < h2 do
        find(t1, [h2|t2], [h1|result_list])
    end

    defp find(l, [h2|t2], result_list) do
        find(l, t2, [h2|result_list])
    end

    def med(l) do
        case rem(Enum.count(l), 2) == 1  do
            true -> Enum.at(l, div(Enum.count(l), 2))
            _ -> ((Enum.at(l, div(Enum.count(l), 2))) + Enum.at(l, div(Enum.count(l) - 1, 2))) / 2
        end
    end
end

File: median.erl
-module(median).
-export([find/2]).

median(L) ->
    case length(L) rem 2 == 1 of
        true -> lists:nth((length(L) div 2) + 1, L);
        false -> (lists:nth((length(L) div 2 + 1), L) + lists:nth(length(L) div 2, L)) / 2
    end.

find(L, []) -> median(L);
find([], L) -> median(L);
find([H1|T1], [H2|T2]) ->
    case H1 < H2 of
        true -> find(T1, [H2|T2], [H1]);
        false -> find([H1|T1], T2, [H2])
    end.

find([], [], ResultList) -> median(ResultList);
find([], [H|T], ResultList) -> find([], T, [H|ResultList]);
find([H|T], [], ResultList) -> find(T, [], [H|ResultList]);
find([H1|T1], [H2|T2], ResultList) when H1 < H2 -> find(T1, [H2|T2], [H1|ResultList]);
find(L, [H2|T2], ResultList) -> find(L, [T2], [H2|ResultList]).
